### PR TITLE
ipn/ipnlocal: [serve] listen on all-interfaces for macOS

### PR DESCRIPTION
**Pull request by shayne**
On macOS (AppStore and macsys), we need to bind to ""/all-interfaces due to the network sandbox. Ideally we would only bind to the Tailscale interface, but that's not possible, macOS fails to listen on privileged ports when binding to a specific interface.

We also implement the lc.Control hook, same as we do for peerapi. It doesn't solve our problem but it's better that we do and would likely be required when Apple gets around to fixing per-interface priviliged port binding.

Fixes: #6364

Signed-off-by: Shayne Sweeney <shayne@tailscale.com>

### CODE REVIEW

File: ipn/ipnlocal/serve.go [hunk 1]
```diff
@@ -91,7 +92,36 @@ func (s *serveListener) Close() error {
...
+			// On iOS/macOS, this sets the lc.Control hook to
+			// setsockopt the interface index to bind to, to get
+			// out of the network sandbox.
+			if err := initListenConfig(&lc, ip, s.b.prevIfState, s.b.dialer.TUNName()); err != nil {
+				s.logf("serve failed to init listen config %v, backing off: %v", s.ap, err)
+				s.bo.BackOff(s.ctx, err)
+				continue
+			}
+			// On macOS and if we're binding to a privileged port,
+			if runtime.GOOS == "darwin" && s.ap.Port() < 1024 {
```
Comment (bradfitz) [inline ipn/ipnlocal/serve.go:109]: there are three mac variants. you only want to do this for 2 of the 3 probably. see the IsSandboxed funcs in the version package.

Comment (shayne) [reply ipn/ipnlocal/serve.go]: Done. Using `version.IsSandboxedMacOS()` which is the 2 of the 3 (AppStore and macsys).

File: ipn/ipnlocal/serve.go [hunk 2]
```diff
@@ -91,7 +92,36 @@ func (s *serveListener) Close() error {
...
 	for {
-		ln, err := net.Listen("tcp", s.ap.String())
+		ip := s.ap.Addr()
+		ipStr := ip.String()
+
+		var lc net.ListenConfig
+		if initListenConfig != nil {
+			// On iOS/macOS, this sets the lc.Control hook to
+			// setsockopt the interface index to bind to, to get
+			// out of the network sandbox.
```
Comment (bradfitz) [inline ipn/ipnlocal/serve.go:102]: can you reword this "to get out of the network sandbox" to explain more what's going on and what the change in behavior is?

File: ipn/ipnlocal/serve.go [hunk 3]
```diff
@@ -91,7 +92,36 @@ func (s *serveListener) Close() error {
 // Listen is retried until the context is canceled.
 func (s *serveListener) Run() {
 	for {
-		ln, err := net.Listen("tcp", s.ap.String())
+		ip := s.ap.Addr()
+		ipStr := ip.String()
+
+		var lc net.ListenConfig
+		if initListenConfig != nil {
+			// On iOS/macOS, this sets the lc.Control hook to
```
Comment (bradfitz) [inline ipn/ipnlocal/serve.go:100]: macOS sandbox only?

File: ipn/ipnlocal/serve.go [hunk 4]
```diff
@@ -91,7 +92,36 @@ func (s *serveListener) Close() error {
...
+				s.bo.BackOff(s.ctx, err)
+				continue
+			}
+			// On macOS (AppStore or macsys) and if we're binding to a privileged port,
+			if version.IsSandboxedMacOS() && s.ap.Port() < 1024 {
+				// On macOS, we need to bind to ""/all-interfaces due to
+				// the network sandbox. Ideally we would only bind to the
+				// Tailscale interface, but macOS errors out if we try to
+				// to listen on privileged ports binding only to a specific
+				// interface. (6364)
```
Comment (bradfitz) [inline ipn/ipnlocal/serve.go:114]: `#6364`

**Review by bradfitz — APPROVED** [review]

**Pull request merged**